### ISSUE TITLE
An predeclared object behaves like an appobject

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#839]
**Describe the bug**
In an ActiveX DLL project, a `PredeclaredId` object gets the `appobject` attribute instead:
![image](https://user-images.githubusercontent.com/2367644/168422082-ef8405da-5b4b-4a2d-ac9b-a080726d8b2e.png)

This is not quite the same thing because one effect of an `appobject` attribute is that all the members of the class are then made available in the global namespace, leading to namespace pollution. When using `PredeclaredId` attribute, it should be required to address the object by its name (e.g. `ArchiveFactory.ArchiveLibPath`, not just `ArchiveLibPath`)

**To Reproduce**
1. Create an ActiveX DLL project
2. Create a class with a public method. Decorate with `PredeclaredId` attribute.
3. Build the project
4. Load the project in a host (e.g. Excel in this case)
5. Observe that Excel's globals now has the object's members listed and the object members can be called without requiring an object prefix.

**Expected behavior**
The object prefix must be mandatory for objects that are decorated with `PredeclaredId` but not `AppObject`. 

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler BETA 38

**Comment by WaynePhillipsEA** [comment]
Indeed, this is the case.  I have a note in the type library builder:
`// FIXME it appears we can't create a class with TYPEFLAG_FPREDECLID without TYPEFLAG_FAPPOBJECT.  This fails on the ICreateTypeInfo::SetTypeFlags call`

So I will have to look into this to figure out why that API is failing.

**Comment by WaynePhillipsEA** [comment]
I just looked into this, and unfortunately it's not possible to create a coclass with the `TYPEFLAG_FPREDECLID` attribute without the `TYPEFLAG_FAPPOBJECT`.  You can't create one with the `CreateTypeInfo` API, and you can't create one using MIDL.

I went further.  I manipulated the TYPEFLAG_FAPPOBJECT bit in the outputted type library (TLB) binary so that I could see if VBx would even respect it if we were allowed to output a coclass with just the TYPEFLAG_FPREDECLID attribute.  The result is that VBx doesn't complain about it, but ultimately it doesn't work.  What happens is that VBx creates the predeclared ID symbol but doesn't auto-populate it, so it throws 'Object required' errors when you try to call members on it.  What you end up with is essentially a global variable of the correct type, but you have to Set-assign it manually, which obviously is not correct.

So we're going to have to disallow creating predeclared objects that are exposed to COM, and make a suggestion/quick-fix to either make them Private / `[ COMCreatable(False) ]` or change them to `[ AppObject ]` (not yet implemented).

**Comment by WaynePhillipsEA** [comment]
Just thinking some more on this.  

We could potentially _simulate_ predeclared objects in COM by exposing them through an appobject wrapper that just exposes them via a property-get.  So to expose `MyPredeclaredClass` the compiler could auto-generate a coclass of `_MyPredeclaredClass_PredeclaredAccessor` (or something like that) that has `[appobject]` set on it, and has only one property-get method... called `MyPredeclaredClass` that simulates the usual predeclared ID behaviour.

**Comment by bclothier** [comment]
I did think it odd that MIDL did not even have an attribute that corresponds to the flag. Some follow up questions to consider:

1) I wonder if that's why some libraries such as Excel or Word make use of `appobject`-decorated `Global` class which in turn exposes more members beyond what is on the `Application`. Not all Office libraries are like that; Access or PowerPoint doesn't have a `Global`; its `Application` is decorated with the `appobject` but either way, it exposes additional objects via those.

2) Did we investigate whether this is allowed in a VB6 Ax DLL? is it able to expose a predeclaredId class without requiring using an `appobject` coclass?

3) It seems unlikely to me but is it a matter of decorating on the interface rather than coclass?

**Comment by WaynePhillipsEA** [comment]
1. Indeed, I would expect so.  In Access for example, `DoCmd` feels very much like a predeclared object, but ultimately hangs off the Application appobject class.

2. No, it is not.  I just double-checked, and it actually manages to create a type library exactly the same as I created by hand... so it does have the TYPFLAG_FPREDECLID set independently, but using the predeclared symbol produces the 'Object required' runtime error due to it not being supported properly in the client (also VBx). 

3. No, `TYPEFLAG_FAPPOBJECT` and `TYPEFLAG_FPREDECLID` cannot be set on interfaces.

**Comment by bclothier** [comment]
Given that we've established this is not something that VB6 was doing (well, it's trying but is broken), we're now in the territory of a feature request rather than matching parity. 

I would agree that creating an `appobject`-decorated coclass and exposing the "predeclaredId" coclasses on it would be the best way forward. We can automate this but for now, manual creation of such `appobject` is probably best.

**Comment by WaynePhillipsEA** [comment]
Stupid me, the `Object required` was due to me testing in the intermediate window where Option Explicit is off.  So actually VBx just ignores the TYPFLAG_FPREDECLID flag if encountered on its own without TYPFLAG_FAPPOBJECT, as simulated by hand.  When creating the DLL from VB6, it just drops (ignores) the Attribute VB_PredeclaredId attribute for generated COM coclasses.

The conclusion is the same though... we can't generate predeclared objects in COM type libraries.

**Comment by WaynePhillipsEA** [comment]
Until this feature is implemented, I would suggest we leave the current behaviour in place (where `[ Predeclared ]` attribute outputs as an appobject).

**Comment by bclothier** [comment]
For future considerations, I suspect we will want to be a bit more smarter about it and consolidate all `PredeclareId` classes into a single `appobject` object. Not sure how to handle the scenarios where the DLL may already have an `appobject` coclass defined (whether to add them to that implicitly or to create a 2nd class). 

Not sure having multiple appobjects in a single type library is desirable.

**Comment by WaynePhillipsEA** [comment]
Yeah, that's out for discussion.  This won't be implemented imminently, so there's time to consider what's best.

**Comment by WaynePhillipsEA** [comment]
Another option might be to instead output the property-getter as a DLL exported function, which could exposed via the generated type library (inside a module type).  That would offer better performance than an appobject wrapper.  Anyway, next...
